From 645d416127f18c5dacaff8945ce4886f67ffb187 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Thu, 2 May 2024 16:51:18 +0200
Subject: [PATCH 0928/1695] drivers: stm32_rifsc: rework RIFSC driver as a
 firewall controller

Use the new firewall API to populate the firewall bus and register
the RIFSC as a firewall provider.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git af4a24e41ba77f4b136399519aa99fa0ac36a088 drivers: stm32_rifsc: rework RIFSC driver as a firewall controller]
Change-Id: I9100a137698b2511d2c6c262d051bc3d1f45942a
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/386452
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 659 +++++++++-------------------
 1 file changed, 214 insertions(+), 445 deletions(-)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index 08950797a..aca38dcff 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -3,15 +3,17 @@
  * Copyright (c) 2021-2023, STMicroelectronics
  */
 
-#include <drivers/stm32_firewall.h>
+#include <drivers/firewall.h>
 #include <drivers/stm32_rif.h>
 #include <drivers/stm32_rifsc.h>
+#include <drivers/stm32_shared_io.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <io.h>
 #include <kernel/boot.h>
 #include <kernel/dt.h>
 #include <kernel/dt_driver.h>
 #include <kernel/panic.h>
+#include <kernel/pm.h>
 #include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <tee_api_defines.h>
@@ -73,7 +75,7 @@
 
 /* Periph id per register */
 #define _PERIPH_IDS_PER_REG		U(32)
-#define _OFST_PERX_CIDCFGR		U(0x8)
+#define _OFFSET_PERX_CIDCFGR		U(0x8)
 
 #define RIFSC_RISC_CFEN_MASK		BIT(0)
 #define RIFSC_RISC_SEM_EN_MASK		BIT(1)
@@ -126,12 +128,6 @@ struct rifsc_platdata {
 	int nrimu;
 };
 
-struct dt_id_attr {
-	/* The effective size of the array is meaningless here */
-	fdt32_t id_attr[1];
-};
-
-static struct stm32_firewall_device *fdev;
 static struct rifsc_driver_data rifsc_drvdata;
 static struct rifsc_platdata rifsc_pdata;
 
@@ -186,8 +182,8 @@ static TEE_Result stm32_rifsc_glock_config(const void *fdt, int node,
 	if (glock_conf & RIFSC_RIMU_GLOCK) {
 		DMSG("Setting global lock on RIMU configuration");
 
-		io_setbits32(pdata->base + _RIFSC_RIMC_CR,
-			     _RIFSC_RIMC_CR_GLOCK);
+		io_setbits32_stm32shregs(pdata->base + _RIFSC_RIMC_CR,
+					 _RIFSC_RIMC_CR_GLOCK);
 
 		if (!(io_read32(pdata->base + _RIFSC_RIMC_CR) &
 		      _RIFSC_RIMC_CR_GLOCK))
@@ -197,7 +193,7 @@ static TEE_Result stm32_rifsc_glock_config(const void *fdt, int node,
 	if (glock_conf & RIFSC_RISUP_GLOCK) {
 		DMSG("Setting global lock on RISUP configuration");
 
-		io_setbits32(pdata->base, _RIFSC_RISC_CR_GLOCK);
+		io_setbits32_stm32shregs(pdata->base, _RIFSC_RISC_CR_GLOCK);
 
 		if (!(io_read32(pdata->base) & _RIFSC_RISC_CR_GLOCK))
 			return TEE_ERROR_ACCESS_DENIED;
@@ -210,13 +206,12 @@ static TEE_Result stm32_rifsc_dt_conf_risup(const void *fdt, int node,
 					    int *nrisup,
 					    struct risup_cfg **risups)
 {
-	const struct dt_id_attr *conf_list = NULL;
+	const fdt32_t *cuint = NULL;
 	int i = 0;
 	int len = 0;
 
-	conf_list = (const struct dt_id_attr *)fdt_getprop(fdt, node,
-							   "st,protreg", &len);
-	if (!conf_list) {
+	cuint = fdt_getprop(fdt, node, "st,protreg", &len);
+	if (!cuint) {
 		DMSG("No RISUP configuration in DT");
 		return TEE_ERROR_ITEM_NOT_FOUND;
 	}
@@ -227,7 +222,7 @@ static TEE_Result stm32_rifsc_dt_conf_risup(const void *fdt, int node,
 		return TEE_ERROR_OUT_OF_MEMORY;
 
 	for (i = 0; i < *nrisup; i++) {
-		uint32_t value = fdt32_to_cpu(conf_list->id_attr[i]);
+		uint32_t value = fdt32_to_cpu(cuint[i]);
 		struct risup_cfg *risup = *risups + i;
 
 		risup->id = _RIF_FLD_GET(RIFSC_RISC_PER_ID, value);
@@ -243,13 +238,12 @@ static TEE_Result stm32_rifsc_dt_conf_risup(const void *fdt, int node,
 static TEE_Result stm32_rifsc_dt_conf_rimu(const void *fdt, int node,
 					   struct rifsc_platdata *pdata)
 {
-	const struct dt_id_attr *conf_list = NULL;
+	const fdt32_t *cuint = NULL;
 	int i = 0;
 	int len = 0;
 
-	conf_list = (const struct dt_id_attr *)fdt_getprop(fdt, node,
-							   "st,rimu", &len);
-	if (!conf_list) {
+	cuint = fdt_getprop(fdt, node, "st,rimu", &len);
+	if (!cuint) {
 		DMSG("No RIMU configuration in DT");
 		return TEE_ERROR_ITEM_NOT_FOUND;
 	}
@@ -262,10 +256,11 @@ static TEE_Result stm32_rifsc_dt_conf_rimu(const void *fdt, int node,
 		return TEE_ERROR_OUT_OF_MEMORY;
 
 	for (i = 0; i < len; i++) {
-		uint32_t value = fdt32_to_cpu(conf_list->id_attr[i]);
+		uint32_t value = fdt32_to_cpu(cuint[i]);
 		struct rimu_cfg *rimu = pdata->rimu + i;
 
-		rimu->id = _RIF_FLD_GET(RIFSC_RIMC_M_ID, value);
+		rimu->id = _RIF_FLD_GET(RIFSC_RIMC_M_ID, value) -
+			   RIMU_ID_OFFSET;
 		rimu->attr = _RIF_FLD_GET(RIFSC_RIMC_ATTRx, value);
 	}
 
@@ -301,109 +296,11 @@ static TEE_Result stm32_rifsc_parse_fdt(const void *fdt, int node,
 	return TEE_SUCCESS;
 }
 
-static TEE_Result
-stm32_rifsc_has_access(struct stm32_firewall_device *dev __unused,
-		       unsigned int id,
-		       paddr_t base __unused, size_t sz __unused,
-		       const struct stm32_firewall_cfg *cfg)
-{
-	TEE_Result res = TEE_ERROR_ACCESS_DENIED;
-	const struct stm32_firewall_cfg *f_cfg = cfg;
-	unsigned int reg_id = id / 32;
-	unsigned int periph_offset = id - (reg_id * 32);
-	uintptr_t rifsc_base = rifsc_pdata.base;
-	uint32_t sec_reg_value = 0;
-	uint32_t priv_reg_value = 0;
-	uint32_t cid_reg_value = 0;
-	uint32_t mode = 0;
-
-	DMSG("Check firewall access for peripheral (ID): %u", id);
-
-	assert(rifsc_base);
-
-	if (!f_cfg)
-		return TEE_ERROR_BAD_PARAMETERS;
-
-	for (f_cfg = cfg; f_cfg->access; f_cfg++) {
-		/* The CID corresponds to the * FWLL_MASTER(ID) + 1 */
-		unsigned int master = ((f_cfg->access & FWLL_MASTER_MASK) >>
-				      FWLL_MASTER_SHIFT) + 1;
-
-		if (master > MAX_CID_SUPPORTED) {
-			EMSG("This CID does not exist: %u", master);
-			panic();
-		}
-
-		mode = cfg->access & FWLL_CONF_MASK;
-
-		if (mode == FWLL_NO_ACCESS)
-			continue;
-
-		sec_reg_value = io_read32(rifsc_base + _RIFSC_RISC_SECCFGR0 +
-					  0x4 * reg_id);
-		priv_reg_value = io_read32(rifsc_base + _RIFSC_RISC_PRIVCFGR0 +
-					   0x4 * reg_id);
-		cid_reg_value = io_read32(rifsc_base +
-					  _RIFSC_RISC_PER0_CIDCFGR +
-					  0x8 * id);
-
-		if (!(mode & (FWLL_SEC_PRIV | FWLL_SEC_UNPRIV |
-			      FWLL_NSEC_PRIV | FWLL_NSEC_UNPRIV)))
-			return TEE_ERROR_BAD_PARAMETERS;
-
-		/*
-		 * First check conditions for semaphore mode, which doesn't
-		 * take into account static CID.
-		 */
-		if (cid_reg_value & _CIDCFGR_SEMEN) {
-			if (cid_reg_value & BIT(master + SEMWL_SHIFT)) {
-				/* Static CID is irrelevant if semaphore mode */
-				goto skip_cid_check;
-			} else {
-				return TEE_ERROR_ACCESS_DENIED;
-			}
-		}
-
-		if (!(cid_reg_value & _CIDCFGR_CFEN) ||
-		    ((cid_reg_value & RIFSC_RISC_SCID_MASK) >>
-		    RIFSC_RISC_SCID_SHIFT) == RIF_CID0)
-			goto skip_cid_check;
-
-		/*
-		 * Coherency check with the CID configuration
-		 */
-		if (((cid_reg_value & RIFSC_RISC_SCID_MASK) >>
-		    RIFSC_RISC_SCID_SHIFT) != master)
-			return TEE_ERROR_ACCESS_DENIED;
-
-skip_cid_check:
-		/* Check non-secure, non-privilege configuration */
-		if (mode & (FWLL_NSEC_UNPRIV) &&
-		    (sec_reg_value & BIT(periph_offset) ||
-		     priv_reg_value & BIT(periph_offset)))
-			return TEE_ERROR_ACCESS_DENIED;
-
-		/* Check secure configuration */
-		if ((mode & (FWLL_SEC_PRIV) || mode & (FWLL_SEC_UNPRIV)) &&
-		    !(sec_reg_value & BIT(periph_offset)))
-			return TEE_ERROR_ACCESS_DENIED;
-
-		/* Check unprivilege configuration */
-		if ((priv_reg_value & BIT(periph_offset)) &&
-		    (!(mode & FWLL_NSEC_PRIV)) && (!(mode & FWLL_SEC_PRIV)))
-			return TEE_ERROR_ACCESS_DENIED;
-
-		res = TEE_SUCCESS;
-	}
-
-	return res;
-}
-
 static TEE_Result stm32_risup_cfg(struct rifsc_platdata *pdata,
 				  struct risup_cfg *risup)
 {
 	struct rifsc_driver_data *drv_data = pdata->drv_data;
-	uintptr_t cidcfgr_offset = _OFST_PERX_CIDCFGR * risup->id;
+	uintptr_t cidcfgr_offset = _OFFSET_PERX_CIDCFGR * risup->id;
 	uintptr_t offset = sizeof(uint32_t) * (risup->id / _PERIPH_IDS_PER_REG);
 	uint32_t shift = risup->id % _PERIPH_IDS_PER_REG;
 	TEE_Result res = TEE_ERROR_GENERIC;
@@ -412,12 +309,14 @@ static TEE_Result stm32_risup_cfg(struct rifsc_platdata *pdata,
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	if (drv_data->sec_en)
-		io_clrsetbits32(pdata->base + _RIFSC_RISC_SECCFGR0 + offset,
-				BIT(shift), risup->sec << shift);
+		io_clrsetbits32_stm32shregs(pdata->base + _RIFSC_RISC_SECCFGR0 +
+					    offset, BIT(shift),
+					    risup->sec << shift);
 
 	if (drv_data->priv_en)
-		io_clrsetbits32(pdata->base + _RIFSC_RISC_PRIVCFGR0 + offset,
-				BIT(shift), risup->priv << shift);
+		io_clrsetbits32_stm32shregs(pdata->base +
+					    _RIFSC_RISC_PRIVCFGR0 + offset,
+					    BIT(shift), risup->priv << shift);
 
 	if (drv_data->rif_en)
 		io_write32(pdata->base + _RIFSC_RISC_PER0_CIDCFGR +
@@ -426,8 +325,8 @@ static TEE_Result stm32_risup_cfg(struct rifsc_platdata *pdata,
 	/* Lock configuration for this RISUP */
 	if (risup->lock) {
 		DMSG("Locking RIF conf for peripheral n°%"PRIu32, risup->id);
-		io_setbits32(pdata->base + _RIFSC_RISC_RCFGLOCKR0 + offset,
-			     BIT(shift));
+		io_setbits32_stm32shregs(pdata->base + _RIFSC_RISC_RCFGLOCKR0 +
+					 offset, BIT(shift));
 	}
 
 	/*
@@ -558,7 +457,7 @@ bool stm32_rifsc_cid_is_enabled(unsigned int  rifc_id)
 	struct io_pa_va rifsc_addr = { .pa = RIFSC_BASE };
 	vaddr_t rifsc_base = io_pa_or_va(&rifsc_addr, 1);
 	uint32_t cidcfgr = io_read32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
-				     _OFST_PERX_CIDCFGR * rifc_id);
+				     _OFFSET_PERX_CIDCFGR * rifc_id);
 
 	return (cidcfgr & _CIDCFGR_CFEN) == _CIDCFGR_CFEN;
 }
@@ -568,8 +467,8 @@ void stm32_rifsc_cid_enable(unsigned int  rifc_id)
 	struct io_pa_va rifsc_addr = { .pa = RIFSC_BASE };
 	vaddr_t rifsc_base = io_pa_or_va(&rifsc_addr, 1);
 
-	io_setbits32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
-		     _OFST_PERX_CIDCFGR * rifc_id, _CIDCFGR_CFEN);
+	io_setbits32_stm32shregs(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
+				 _OFFSET_PERX_CIDCFGR * rifc_id, _CIDCFGR_CFEN);
 }
 
 void stm32_rifsc_cid_disable(unsigned int  rifc_id)
@@ -577,48 +476,186 @@ void stm32_rifsc_cid_disable(unsigned int  rifc_id)
 	struct io_pa_va rifsc_addr = { .pa = RIFSC_BASE };
 	vaddr_t rifsc_base = io_pa_or_va(&rifsc_addr, 1);
 
-	io_clrbits32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
-		     _OFST_PERX_CIDCFGR * rifc_id, _CIDCFGR_CFEN);
+	io_clrbits32_stm32shregs(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
+				 _OFFSET_PERX_CIDCFGR * rifc_id, _CIDCFGR_CFEN);
 }
 
-TEE_Result stm32_rifsc_reconfigure_rimu(unsigned int id,
-					unsigned int master_cid,
-					bool cid_sel, bool sec, bool priv)
+static TEE_Result stm32_rifsc_check_access(struct firewall_query *firewall)
 {
-	struct rimu_cfg *rimu = NULL;
-	uint32_t rimu_conf = 0;
-	TEE_Result res = TEE_ERROR_GENERIC;
+	uintptr_t rifsc_base = rifsc_pdata.base;
+	unsigned int cid_reg_offset = 0;
+	unsigned int periph_offset = 0;
+	uint32_t cid_to_check = 0;
+	unsigned int reg_id = 0;
+	bool priv_check = true;
+	bool sec_check = true;
+	uint32_t privcfgr = 0;
+	uint32_t seccfgr = 0;
+	uint32_t cidcfgr = 0;
+
+	assert(rifsc_base);
 
-	if (id > rifsc_pdata.drv_data->nb_rimu ||
-	    master_cid > MAX_CID_SUPPORTED)
+	if (!firewall || firewall->arg_count != 2)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	rimu = &rifsc_pdata.rimu[id];
+	/*
+	 * Peripheral configuration, we assume the configuration is as
+	 * follows:
+	 * firewall->args[0]: Firewall ID
+	 * firewall->args[1]: RIF configuration to check, reusing RIFPROT()
+	 *		      macro. Only secure, privilege and static CID
+	 *		      fields are used
+	 */
+	if (firewall->args[0] >= RIMU_ID_OFFSET)
+		return TEE_SUCCESS;
+
+	reg_id = firewall->args[0] / _PERIPH_IDS_PER_REG;
+	periph_offset = firewall->args[0] % _PERIPH_IDS_PER_REG;
+	cid_reg_offset = _OFFSET_PERX_CIDCFGR * firewall->args[0];
+	cidcfgr = io_read32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
+			    cid_reg_offset);
+	seccfgr = io_read32(rifsc_base + _RIFSC_RISC_SECCFGR0 + 0x4 * reg_id);
+	privcfgr = io_read32(rifsc_base + _RIFSC_RISC_PRIVCFGR0 + 0x4 * reg_id);
+	sec_check = _RIF_FLD_GET(RIFSC_RISC_SEC, firewall->args[1]) != 0;
+	priv_check = _RIF_FLD_GET(RIFSC_RISC_PRIV, firewall->args[1]) != 0;
+	cid_to_check = _RIF_FLD_GET(RIFSC_RISC_PERx_CID, firewall->args[1]) >>
+		       RIFSC_RISC_SCID_SHIFT;
+
+	if (!sec_check && seccfgr & BIT(periph_offset))
+		return TEE_ERROR_ACCESS_DENIED;
 
-	if (io_read32(rifsc_pdata.base + _RIFSC_RIMC_CR) &
-	    _RIFSC_RIMC_CR_GLOCK) {
-		DMSG("RIMU configuration is locked");
+	if (!priv_check && (privcfgr & BIT(periph_offset)))
 		return TEE_ERROR_ACCESS_DENIED;
+
+	if (!(cidcfgr & _CIDCFGR_CFEN))
+		return TEE_SUCCESS;
+
+	if ((cidcfgr & _CIDCFGR_SEMEN &&
+	     !SEM_EN_AND_OK(cidcfgr, cid_to_check)) ||
+	    (!(cidcfgr & _CIDCFGR_SEMEN) &&
+	     !SCID_OK(cidcfgr, RIFSC_RISC_SCID_MASK, cid_to_check)))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_rifsc_acquire_access(struct firewall_query *firewall)
+{
+	uintptr_t rifsc_base = rifsc_pdata.base;
+	unsigned int cid_reg_offset = 0;
+	uint32_t cidcfgr = 0;
+
+	assert(rifsc_base);
+
+	if (!firewall || !firewall->arg_count)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (firewall->args[0] >= RIMU_ID_OFFSET)
+		return TEE_SUCCESS;
+
+	cid_reg_offset = _OFFSET_PERX_CIDCFGR * firewall->args[0];
+	cidcfgr = io_read32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
+			    cid_reg_offset);
+
+	FMSG("CID filtering %s ,static CID%u ,semaphore list:%#x for periph %u",
+	     (cidcfgr & _CIDCFGR_CFEN) ? "enabled" : "disabled",
+	     (cidcfgr & RIFSC_RISC_SCID_MASK) >> SCID_SHIFT,
+	     (cidcfgr & RIFSC_RISC_SEML_MASK) >> SEMWL_SHIFT,
+	     firewall->args[0]);
+
+	/* Only check CID attributes */
+	if (!(cidcfgr & _CIDCFGR_CFEN))
+		return TEE_SUCCESS;
+
+	if (cidcfgr & _CIDCFGR_SEMEN) {
+		if (!SEM_EN_AND_OK(cidcfgr, RIF_CID1))
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		/* Take the semaphore, static CID is irrelevant here */
+		return stm32_rif_acquire_semaphore(rifsc_base +
+						   _RIFSC_RISC_PER0_SEMCR +
+						   cid_reg_offset,
+						   MAX_CID_SUPPORTED);
 	}
 
-	if (cid_sel)
-		rimu_conf |= RIFSC_RIMC_MODE_MASK;
-	if (sec)
-		rimu_conf |= RIFSC_RIMC_MSEC_MASK;
-	if (priv)
-		rimu_conf |= RIFSC_RIMC_MPRIV_MASK;
+	if (!SCID_OK(cidcfgr, RIFSC_RISC_SCID_MASK, RIF_CID1))
+		return TEE_ERROR_ACCESS_DENIED;
 
-	rimu_conf |= master_cid << RIFSC_RIMC_MCID_SHIFT;
+	return TEE_SUCCESS;
+}
 
-	rimu->attr = rimu_conf;
+static TEE_Result stm32_rifsc_set_config(struct firewall_query *firewall)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct rimu_cfg rimu = { };
+	bool is_tdcid = false;
+	unsigned int id = 0;
+	uint32_t conf = 0;
 
-	res = stm32_rimu_cfg(&rifsc_pdata, rimu);
-	if (res) {
-		EMSG("RIMU %u reconfiguration error", id);
+	res = stm32_rifsc_check_tdcid(&is_tdcid);
+	if (res)
 		return res;
+
+	if (!firewall || firewall->arg_count != 2)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	id = firewall->args[0];
+	conf = firewall->args[1];
+
+	if (id < RIMU_ID_OFFSET) {
+		struct risup_cfg risup = { };
+		uint32_t cidcfgr = 0;
+
+		risup.id = firewall->args[0];
+		risup.sec = _RIF_FLD_GET(RIFSC_RISC_SEC, conf) != 0;
+		risup.priv = _RIF_FLD_GET(RIFSC_RISC_PRIV, conf) != 0;
+		risup.lock = _RIF_FLD_GET(RIFSC_RISC_LOCK, conf) != 0;
+		risup.cid_attr = _RIF_FLD_GET(RIFSC_RISC_PERx_CID, conf);
+
+		if (!is_tdcid) {
+			cidcfgr = io_read32(rifsc_pdata.base +
+					    _OFFSET_PERX_CIDCFGR * risup.id +
+					    _RIFSC_RISC_PER0_CIDCFGR);
+
+			if (cidcfgr != risup.cid_attr)
+				return TEE_ERROR_BAD_PARAMETERS;
+		}
+
+		return stm32_risup_cfg(&rifsc_pdata, &risup);
 	}
 
-	return TEE_SUCCESS;
+	if (!is_tdcid)
+		return TEE_ERROR_ACCESS_DENIED;
+
+	rimu.id = _RIF_FLD_GET(RIFSC_RIMC_M_ID, conf) - RIMU_ID_OFFSET;
+	rimu.attr = _RIF_FLD_GET(RIFSC_RIMC_ATTRx, conf);
+
+	return stm32_rimu_cfg(&rifsc_pdata, &rimu);
+}
+
+static void stm32_rifsc_release_access(struct firewall_query *firewall)
+{
+	uintptr_t rifsc_base = rifsc_pdata.base;
+	uint32_t cidcfgr = 0;
+	uint32_t id = 0;
+
+	assert(rifsc_base && firewall && firewall->arg_count);
+
+	id = firewall->args[0];
+
+	if (id >= RIMU_ID_OFFSET)
+		return;
+
+	cidcfgr = io_read32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
+			    _OFFSET_PERX_CIDCFGR * id);
+
+	/* Only thing possible is to release a semaphore taken by OP-TEE CID */
+	if (SEM_EN_AND_OK(cidcfgr, RIF_CID1))
+		if (stm32_rif_release_semaphore(rifsc_base +
+						_RIFSC_RISC_PER0_SEMCR +
+						id * _OFFSET_PERX_CIDCFGR,
+						MAX_CID_SUPPORTED))
+			panic("Could not release the RIF semaphore");
 }
 
 static TEE_Result stm32_rifsc_sem_pm_suspend(void)
@@ -628,7 +665,7 @@ static TEE_Result stm32_rifsc_sem_pm_suspend(void)
 	for (i = 0; i < rifsc_pdata.nrisup && i < rifsc_drvdata.nb_risup; i++) {
 		uint32_t semcfgr = io_read32(rifsc_pdata.base +
 					     _RIFSC_RISC_PER0_SEMCR +
-					     _OFST_PERX_CIDCFGR * i);
+					     _OFFSET_PERX_CIDCFGR * i);
 		struct risup_cfg *risup = rifsc_pdata.risup + i;
 
 		/* Save semaphores that were taken by the CID1 */
@@ -651,7 +688,7 @@ static TEE_Result stm32_rifsc_sem_pm_resume(void)
 
 	for (i = 0; i < rifsc_pdata.nrisup && i < rifsc_drvdata.nb_risup; i++) {
 		struct risup_cfg *risup = rifsc_pdata.risup + i;
-		uintptr_t cidcfgr_offset = _OFST_PERX_CIDCFGR * risup->id;
+		uintptr_t cidcfgr_offset = _OFFSET_PERX_CIDCFGR * risup->id;
 		uintptr_t offset = sizeof(uint32_t) *
 				   (risup->id / _PERIPH_IDS_PER_REG);
 		uintptr_t perih_offset = risup->id % _PERIPH_IDS_PER_REG;
@@ -719,56 +756,17 @@ TEE_Result stm32_rifsc_check_tdcid(bool *tdcid_state)
 	return TEE_SUCCESS;
 }
 
-static TEE_Result
-stm32_rifsc_get_data_config(const void *fdt, int node,
-			    int *ndata, void **data_cfg)
-{
-	if (stm32_rifsc_dt_conf_risup(fdt, node, ndata,
-				      (struct risup_cfg **)data_cfg))
-		return TEE_ERROR_ITEM_NOT_FOUND;
-
-	return TEE_SUCCESS;
-}
-
-static TEE_Result
-stm32_rifsc_set_data_config(int ndata, void *data_cfg)
-{
-	struct rifsc_platdata *pdata = &rifsc_pdata;
-	struct rifsc_driver_data *drv_data = pdata->drv_data;
-	int i = 0;
-	TEE_Result res = TEE_ERROR_GENERIC;
-
-	for (i = 0; i < ndata && i < drv_data->nb_risup; i++) {
-		struct risup_cfg *risup = (struct risup_cfg *)data_cfg + i;
-
-		res = stm32_risup_cfg(pdata, risup);
-		if (res) {
-			EMSG("risup cfg(%d/%d) error", i + 1, ndata);
-			return res;
-		}
-	}
-
-	return TEE_SUCCESS;
-}
-
-static TEE_Result
-stm32_rifsc_put_data_config(int ndata __unused, void *data_cfg)
-{
-	free(data_cfg);
-
-	return TEE_SUCCESS;
-}
-
-static const struct stm32_firewall_ops stm32_rifsc_fw_ops = {
-	.has_access = stm32_rifsc_has_access,
-	.get_data_config = stm32_rifsc_get_data_config,
-	.set_data_config = stm32_rifsc_set_data_config,
-	.put_data_config = stm32_rifsc_put_data_config,
+static const struct firewall_controller_ops firewall_ops = {
+	.set_conf = stm32_rifsc_set_config,
+	.check_access = stm32_rifsc_check_access,
+	.acquire_access = stm32_rifsc_acquire_access,
+	.release_access = stm32_rifsc_release_access,
 };
 
 static TEE_Result stm32_rifsc_probe(const void *fdt, int node,
-				    const void *compat_data)
+				    const void *compat_data __unused)
 {
+	struct firewall_controller *controller = NULL;
 	TEE_Result res = TEE_ERROR_GENERIC;
 
 	res = stm32_rifsc_parse_fdt(fdt, node, &rifsc_pdata);
@@ -788,260 +786,31 @@ static TEE_Result stm32_rifsc_probe(const void *fdt, int node,
 
 	stm32_rifsc_glock_config(fdt, node, &rifsc_pdata);
 
-	fdev = stm32_firewall_dev_alloc();
-	if (!fdev)
-		return TEE_ERROR_OUT_OF_MEMORY;
+	controller = calloc(1, sizeof(*controller));
+	if (!controller)
+		panic();
 
-	fdev->name = "rifsc";
-	fdev->ops = &stm32_rifsc_fw_ops;
-	fdev->compat = (struct stm32_firewall_compat *)compat_data;
-	fdev->priv = (void *)&rifsc_pdata;
+	controller->name = "RIFSC";
+	controller->priv = &rifsc_pdata;
+	controller->ops = &firewall_ops;
 
-	res = stm32_firewall_dev_register(fdev);
+	res = firewall_dt_controller_register(fdt, node, controller);
 	if (res)
-		goto out;
+		panic();
 
-	stm32_firewall_bus_probe(fdev, fdt, node);
+	res = firewall_dt_probe_bus(fdt, node, controller);
+	if (res)
+		panic();
 
 	register_pm_core_service_cb(stm32_rifsc_sem_pm, NULL,
 				    "stm32-rifsc-semaphores");
 
-out:
-	if (res)
-		free(fdev);
-
-	return res;
+	return TEE_SUCCESS;
 }
 
-#if (defined(CFG_STM32MP25) || defined(CFG_STM32MP23))
-static struct stm32_firewall_reg
-rifsc_compat_ids[STM32MP25_RIFSC_IAC_ID + 1] = {
-	[STM32MP25_RIFSC_TIM1_ID] = { 0x40200000 },
-	[STM32MP25_RIFSC_TIM2_ID] = { 0x40000000 },
-	[STM32MP25_RIFSC_TIM3_ID] = { 0x40010000 },
-	[STM32MP25_RIFSC_TIM4_ID] = { 0x40020000 },
-	[STM32MP25_RIFSC_TIM5_ID] = { 0x40030000 },
-	[STM32MP25_RIFSC_TIM6_ID] = { 0x40040000 },
-	[STM32MP25_RIFSC_TIM7_ID] = { 0x40050000 },
-	[STM32MP25_RIFSC_TIM8_ID] = { 0x40210000 },
-	[STM32MP25_RIFSC_TIM10_ID] = { 0x401C0000 },
-	[STM32MP25_RIFSC_TIM11_ID] = { 0x401D0000 },
-	[STM32MP25_RIFSC_TIM12_ID] = { 0x40060000 },
-	[STM32MP25_RIFSC_TIM13_ID] = { 0x40070000 },
-	[STM32MP25_RIFSC_TIM14_ID] = { 0x40080000 },
-	[STM32MP25_RIFSC_TIM15_ID] = { 0x40250000 },
-	[STM32MP25_RIFSC_TIM16_ID] = { 0x40260000 },
-	[STM32MP25_RIFSC_TIM17_ID] = { 0x40270000 },
-	[STM32MP25_RIFSC_TIM20_ID] = { 0x40320000 },
-	[STM32MP25_RIFSC_LPTIM1_ID] = { 0x40090000 },
-	[STM32MP25_RIFSC_LPTIM2_ID] = { 0x400A0000 },
-	[STM32MP25_RIFSC_LPTIM3_ID] = { 0x46050000 },
-	[STM32MP25_RIFSC_LPTIM4_ID] = { 0x46060000 },
-	[STM32MP25_RIFSC_LPTIM5_ID] = { 0x46070000 },
-	[STM32MP25_RIFSC_SPI1_ID] = { 0x40230000 },
-	[STM32MP25_RIFSC_SPI2_ID] = { 0x400B0000 },
-	[STM32MP25_RIFSC_SPI3_ID] = { 0x400C0000 },
-	[STM32MP25_RIFSC_SPI4_ID] = { 0x40240000 },
-	[STM32MP25_RIFSC_SPI5_ID] = { 0x40280000 },
-	[STM32MP25_RIFSC_SPI6_ID] = { 0x40350000 },
-	[STM32MP25_RIFSC_SPI7_ID] = { 0x40360000 },
-	[STM32MP25_RIFSC_SPI8_ID] = { 0x46020000 },
-	[STM32MP25_RIFSC_SPDIFRX_ID] = { 0x400D0000 },
-	[STM32MP25_RIFSC_USART1_ID] = { 0x40330000 },
-	[STM32MP25_RIFSC_USART2_ID] = { 0x400E0000 },
-	[STM32MP25_RIFSC_USART3_ID] = { 0x400F0000 },
-	[STM32MP25_RIFSC_UART4_ID] = { 0x40100000 },
-	[STM32MP25_RIFSC_UART5_ID] = { 0x40110000 },
-	[STM32MP25_RIFSC_USART6_ID] = { 0x40220000 },
-	[STM32MP25_RIFSC_UART7_ID] = { 0x40370000 },
-	[STM32MP25_RIFSC_UART8_ID] = { 0x40380000 },
-	[STM32MP25_RIFSC_UART9_ID] = { 0x402C0000 },
-	[STM32MP25_RIFSC_LPUART1_ID] = { 0x46030000 },
-	[STM32MP25_RIFSC_I2C1_ID] = { 0x40120000 },
-	[STM32MP25_RIFSC_I2C2_ID] = { 0x40130000 },
-	[STM32MP25_RIFSC_I2C3_ID] = { 0x40140000 },
-	[STM32MP25_RIFSC_I2C4_ID] = { 0x40150000 },
-	[STM32MP25_RIFSC_I2C5_ID] = { 0x40160000 },
-	[STM32MP25_RIFSC_I2C6_ID] = { 0x40170000 },
-	[STM32MP25_RIFSC_I2C7_ID] = { 0x40180000 },
-	[STM32MP25_RIFSC_I2C8_ID] = { 0x46040000 },
-	[STM32MP25_RIFSC_SAI1_ID] = { 0x40290000 },
-	[STM32MP25_RIFSC_SAI2_ID] = { 0x402A0000 },
-	[STM32MP25_RIFSC_SAI3_ID] = { 0x402B0000 },
-	[STM32MP25_RIFSC_SAI4_ID] = { 0x40340000},
-	[STM32MP25_RIFSC_MDF1_ID] = { 0x404D0000 },
-	[STM32MP25_RIFSC_ADF1_ID] = { 0x46220000 },
-	[STM32MP25_RIFSC_FDCAN_ID] = { 0xFFFFFFFF },
-	[STM32MP25_RIFSC_HDP_ID] = { 0x44090000 },
-	[STM32MP25_RIFSC_ADC12_ID] = { 0x404E0000 },
-	[STM32MP25_RIFSC_ADC3_ID] = { 0x404F0000 },
-	[STM32MP25_RIFSC_ETH1_ID] = { 0x482C0000 },
-	[STM32MP25_RIFSC_ETH2_ID] = { 0x482D0000 },
-	[STM32MP25_RIFSC_USBH_ID] = { 0xFFFFFFFF },
-	[STM32MP25_RIFSC_USB3DR_ID] = { 0x48300000 },
-	[STM32MP25_RIFSC_COMBOPHY_ID] = { 0x480C0000 },
-	[STM32MP25_RIFSC_PCIE_ID] = { 0x48400000 },
-	[STM32MP25_RIFSC_UCPD1_ID] = { 0x480A0000 },
-	[STM32MP25_RIFSC_ETHSW_DEIP_ID] = { 0x4C000000 },
-	[STM32MP25_RIFSC_ETHSW_ACM_CFG_ID] = { 0x4B080000 },
-	[STM32MP25_RIFSC_ETHSW_ACM_MSGBUF_ID] = { 0x4B000000 },
-	[STM32MP25_RIFSC_STGEN_ID] = { 0x48080000 },
-	[STM32MP25_RIFSC_OCTOSPI1_ID] = { 0x40430000 },
-	[STM32MP25_RIFSC_OCTOSPI2_ID] = { 0x40440000 },
-	[STM32MP25_RIFSC_SDMMC1_ID] = { 0x48220000 },
-	[STM32MP25_RIFSC_SDMMC2_ID] = { 0x48230000 },
-	[STM32MP25_RIFSC_SDMMC3_ID] = { 0x48240000 },
-	[STM32MP25_RIFSC_GPU_ID] = { 0x48280000 },
-	[STM32MP25_RIFSC_LTDC_CMN_ID] = { 0x48010000 },
-	[STM32MP25_RIFSC_DSI_CMN_ID] = { 0x48000000 },
-	[STM32MP25_RIFSC_LVDS_ID] = { 0x48060000 },
-	[STM32MP25_RIFSC_CSI_ID] = { 0x48020000 },
-	[STM32MP25_RIFSC_DCMIPP_ID] = { 0x48030000 },
-	[STM32MP25_RIFSC_DCMI_PSSI_ID] = { 0xFFFFFFFF },
-	[STM32MP25_RIFSC_VDEC_ID] = { 0x480D0000 },
-	[STM32MP25_RIFSC_VENC_ID] = { 0x480E0000 },
-	[STM32MP25_RIFSC_RNG_ID] = { 0x42020000 },
-	[STM32MP25_RIFSC_PKA_ID] = { 0x42060000 },
-	[STM32MP25_RIFSC_SAES_ID] = { 0x42050000 },
-	[STM32MP25_RIFSC_HASH_ID] = { 0x42010000 },
-	[STM32MP25_RIFSC_CRYP1_ID] = { 0x42030000 },
-	[STM32MP25_RIFSC_CRYP2_ID] = { 0x42040000 },
-	[STM32MP25_RIFSC_IWDG1_ID] = { 0x44010000 },
-	[STM32MP25_RIFSC_IWDG2_ID] = { 0x44020000 },
-	[STM32MP25_RIFSC_IWDG3_ID] = { 0x44030000 },
-	[STM32MP25_RIFSC_IWDG4_ID] = { 0x44040000 },
-	[STM32MP25_RIFSC_IWDG5_ID] = { 0x46090000 },
-	[STM32MP25_RIFSC_WWDG1_ID] = { 0x44050000 },
-	[STM32MP25_RIFSC_WWDG2_ID] = { 0x460A0000 },
-	[STM32MP25_RIFSC_VREFBUF_ID] = { 0x44060000 },
-	[STM32MP25_RIFSC_DTS_ID] = { 0x44070000 },
-	[STM32MP25_RIFSC_RAMCFG_ID] = { 0x42070000 },
-	[STM32MP25_RIFSC_CRC_ID] = { 0x404C0000 },
-	[STM32MP25_RIFSC_SERC_ID] = { 0x44080000 },
-	[STM32MP25_RIFSC_OCTOSPIM_ID] = { 0x40500000 },
-	[STM32MP25_RIFSC_GICV2M_ID] = { 0x48090000 },
-	[STM32MP25_RIFSC_I3C1_ID] = { 0x40190000 },
-	[STM32MP25_RIFSC_I3C2_ID] = { 0x401A0000 },
-	[STM32MP25_RIFSC_I3C3_ID] = { 0x401B0000 },
-	[STM32MP25_RIFSC_I3C4_ID] = { 0x46080000 },
-	[STM32MP25_RIFSC_ICACHE_DCACHE_ID] = { 0x40470000 },
-	[STM32MP25_RIFSC_LTDC_L0L1_ID] = { 0xFFFFFFFF },
-	[STM32MP25_RIFSC_LTDC_L2_ID] = { 0xFFFFFFFF },
-	[STM32MP25_RIFSC_LTDC_ROT_ID] = { 0xFFFFFFFF },
-	[STM32MP25_RIFSC_DSI_TRIG_ID] = { 0xFFFFFFFF },
-	[STM32MP25_RIFSC_DSI_RDFIFO_ID] = { 0xFFFFFFFF },
-	[STM32MP25_RIFSC_OTFDEC1_ID] = { 0x40450000 },
-	[STM32MP25_RIFSC_OTFDEC2_ID] = { 0x40460000 },
-	[STM32MP25_RIFSC_IAC_ID] = { 0x42090000 },
-};
-#endif /* CFG_STM32MP25 */
-
-#if defined(CFG_STM32MP21)
-static struct stm32_firewall_reg
-rifsc_compat_ids[STM32MP21_RIFSC_IAC_ID + 1] = {
-	[STM32MP21_RIFSC_TIM1_ID] = { 0x40200000 },
-	[STM32MP21_RIFSC_TIM2_ID] = { 0x40000000 },
-	[STM32MP21_RIFSC_TIM3_ID] = { 0x40010000 },
-	[STM32MP21_RIFSC_TIM4_ID] = { 0x40020000 },
-	[STM32MP21_RIFSC_TIM5_ID] = { 0x40030000 },
-	[STM32MP21_RIFSC_TIM6_ID] = { 0x40040000 },
-	[STM32MP21_RIFSC_TIM7_ID] = { 0x40050000 },
-	[STM32MP21_RIFSC_TIM8_ID] = { 0x40210000 },
-	[STM32MP21_RIFSC_TIM10_ID] = { 0x401C0000 },
-	[STM32MP21_RIFSC_TIM11_ID] = { 0x401D0000 },
-	[STM32MP21_RIFSC_TIM12_ID] = { 0x40060000 },
-	[STM32MP21_RIFSC_TIM13_ID] = { 0x40070000 },
-	[STM32MP21_RIFSC_TIM14_ID] = { 0x40080000 },
-	[STM32MP21_RIFSC_TIM15_ID] = { 0x40250000 },
-	[STM32MP21_RIFSC_TIM16_ID] = { 0x40260000 },
-	[STM32MP21_RIFSC_TIM17_ID] = { 0x40270000 },
-	[STM32MP21_RIFSC_LPTIM1_ID] = { 0x40090000 },
-	[STM32MP21_RIFSC_LPTIM2_ID] = { 0x400A0000 },
-	[STM32MP21_RIFSC_LPTIM3_ID] = { 0x46050000 },
-	[STM32MP21_RIFSC_LPTIM4_ID] = { 0x46060000 },
-	[STM32MP21_RIFSC_LPTIM5_ID] = { 0x46070000 },
-	[STM32MP21_RIFSC_SPI1_ID] = { 0x40230000 },
-	[STM32MP21_RIFSC_SPI2_ID] = { 0x400B0000 },
-	[STM32MP21_RIFSC_SPI3_ID] = { 0x400C0000 },
-	[STM32MP21_RIFSC_SPI4_ID] = { 0x40240000 },
-	[STM32MP21_RIFSC_SPI5_ID] = { 0x40280000 },
-	[STM32MP21_RIFSC_SPI6_ID] = { 0x40350000 },
-	[STM32MP21_RIFSC_SPDIFRX_ID] = { 0x400D0000 },
-	[STM32MP21_RIFSC_USART1_ID] = { 0x40330000 },
-	[STM32MP21_RIFSC_USART2_ID] = { 0x400E0000 },
-	[STM32MP21_RIFSC_USART3_ID] = { 0x400F0000 },
-	[STM32MP21_RIFSC_UART4_ID] = { 0x40100000 },
-	[STM32MP21_RIFSC_UART5_ID] = { 0x40110000 },
-	[STM32MP21_RIFSC_USART6_ID] = { 0x40220000 },
-	[STM32MP21_RIFSC_UART7_ID] = { 0x40370000 },
-	[STM32MP21_RIFSC_LPUART1_ID] = { 0x46030000 },
-	[STM32MP21_RIFSC_I2C1_ID] = { 0x40170000 },
-	[STM32MP21_RIFSC_I2C2_ID] = { 0x40180000 },
-	[STM32MP21_RIFSC_I2C3_ID] = { 0x46040000 },
-	[STM32MP21_RIFSC_SAI1_ID] = { 0x40290000 },
-	[STM32MP21_RIFSC_SAI2_ID] = { 0x402A0000 },
-	[STM32MP21_RIFSC_SAI3_ID] = { 0x402B0000 },
-	[STM32MP21_RIFSC_SAI4_ID] = { 0x40340000},
-	[STM32MP21_RIFSC_MDF1_ID] = { 0x404D0000 },
-	[STM32MP21_RIFSC_FDCAN_ID] = { 0xFFFFFFFF },
-	[STM32MP21_RIFSC_HDP_ID] = { 0x44090000 },
-	[STM32MP21_RIFSC_ADC1_ID] = { 0x404E0000 },
-	[STM32MP21_RIFSC_ADC2_ID] = { 0x404F0000 },
-	[STM32MP21_RIFSC_ETH1_ID] = { 0x482C0000 },
-	[STM32MP21_RIFSC_ETH2_ID] = { 0x482D0000 },
-	[STM32MP21_RIFSC_USBH_ID] = { 0xFFFFFFFF },
-	[STM32MP21_RIFSC_OTG_HS_ID] = { 0x48300000 },
-	[STM32MP21_RIFSC_STGEN_ID] = { 0x48080000 },
-	[STM32MP21_RIFSC_OCTOSPI1_ID] = { 0x40430000 },
-	[STM32MP21_RIFSC_SDMMC1_ID] = { 0x48220000 },
-	[STM32MP21_RIFSC_SDMMC2_ID] = { 0x48230000 },
-	[STM32MP21_RIFSC_SDMMC3_ID] = { 0x48240000 },
-	[STM32MP21_RIFSC_LTDC_CMN_ID] = { 0x48010000 },
-	[STM32MP21_RIFSC_CSI_ID] = { 0x48020000 },
-	[STM32MP21_RIFSC_DCMIPP_ID] = { 0x48030000 },
-	[STM32MP21_RIFSC_DCMI_PSSI_ID] = { 0xFFFFFFFF },
-	[STM32MP21_RIFSC_CCB_ID] = { 0x42037C00 },
-	[STM32MP21_RIFSC_RNG1_ID] = { 0x42030800 },
-	[STM32MP21_RIFSC_RNG2_ID] = { 0x42020000 },
-	[STM32MP21_RIFSC_PKA_ID] = { 0x42032000 },
-	[STM32MP21_RIFSC_SAES_ID] = { 0x42030c00 },
-	[STM32MP21_RIFSC_HASH1_ID] = { 0x42030400 },
-	[STM32MP21_RIFSC_HASH2_ID] = { 0x42010000 },
-	[STM32MP21_RIFSC_CRYP1_ID] = { 0x42030000 },
-	[STM32MP21_RIFSC_CRYP2_ID] = { 0x42040000 },
-	[STM32MP21_RIFSC_IWDG1_ID] = { 0x44010000 },
-	[STM32MP21_RIFSC_IWDG2_ID] = { 0x44020000 },
-	[STM32MP21_RIFSC_IWDG3_ID] = { 0x44030000 },
-	[STM32MP21_RIFSC_IWDG4_ID] = { 0x44040000 },
-	[STM32MP21_RIFSC_WWDG1_ID] = { 0x44050000 },
-	[STM32MP21_RIFSC_VREFBUF_ID] = { 0x44060000 },
-	[STM32MP21_RIFSC_DTS_ID] = { 0x44070000 },
-	[STM32MP21_RIFSC_RAMCFG_ID] = { 0x42070000 },
-	[STM32MP21_RIFSC_CRC_ID] = { 0x404C0000 },
-	[STM32MP21_RIFSC_SERC_ID] = { 0x44080000 },
-	[STM32MP21_RIFSC_I3C1_ID] = { 0x40190000 },
-	[STM32MP21_RIFSC_I3C2_ID] = { 0x401A0000 },
-	[STM32MP21_RIFSC_I3C3_ID] = { 0x401B0000 },
-	[STM32MP21_RIFSC_ICACHE_DCACHE_ID] = { 0x40470000 },
-	[STM32MP21_RIFSC_LTDC_L0L1_ID] = { 0xFFFFFFFF },
-	[STM32MP21_RIFSC_LTDC_L2_ID] = { 0xFFFFFFFF },
-	[STM32MP21_RIFSC_LTDC_ROT_ID] = { 0xFFFFFFFF },
-	[STM32MP21_RIFSC_OTFDEC1_ID] = { 0x40450000 },
-	[STM32MP21_RIFSC_IAC_ID] = { 0x42090000 },
-};
-#endif /* CFG_STM32MP21 */
-
-static const struct stm32_firewall_compat rifsc_compat[] = {
-	{
-		.reg = rifsc_compat_ids,
-		.compat_size = ARRAY_SIZE(rifsc_compat_ids)
-	},
-};
-
 static const struct dt_device_match rifsc_match_table[] = {
 	{
 		.compatible = "st,stm32mp25-rifsc",
-		.compat_data = (void *)rifsc_compat,
 	},
 	{ }
 };
-- 
2.47.2

